keps/sig-cli/kubectl-commands-in-headers.md

```diff
@@ -0,0 +1,208 @@
+- [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Release Signoff Checklist
+
+- [x] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+  - [#859](https://github.com/kubernetes/enhancements/issues/859)
+- [x] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+  - Standard Unit and Integration testing should be sufficient
+- [x] Graduation criteria is in place
+  - This is not a user facing API change
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Requests sent to the apiserver from kubectl include http request headers with context about the kubectl command that
+created the request.  This information could be used by cluster administrators for debugging or
+to gather telemetry about how users are interacting with the cluster.
+
+
+## Motivation
+
+Kubectl generates requests sent to the apiserver for commands such as `apply`, `delete`, `edit`, `run`, however
+the context of the command for the requests is lost and unavailable to cluster administrators.  Context would be
+useful to cluster admins both for debugging the cause of requests as well as providing telemetry about how users
+are interacting with the cluster.
+
+### Goals
+
+- Allow cluster administrators to identify how requests in the logs were generated from
+  kubectl commands.
+
+### Non-Goals
+
+### Anti-Goals
+
+*We explicitly don't want the following*
+
+- Make decisions of any sort in the apiserver based on these headers.
+  - This information is intended to be used by humans-only, specifically **for debugging and telemetry**.
+
+## Proposal
+
+Include in requests made from kubectl to the apiserver:
+
+- the kubectl subcommand
+- which flags were specified as well as whitelisted enum values for flags (never arbitrary values)
+- a generated session id
+- never include the flag values directly, only use a predefined enumeration
+- never include arguments to the commands
+
+### Kubectl-Command Header
+
+The `Kubectl-Command` Header contains the kubectl sub command.  It contains
```

Comment:
Aren’t we supposed to put X- in front of custom headers,

Reply:
My understanding is that is deprecated: [link](https://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions)

Reply:
I think the deprecation is primarily for headers that may be standardized in an RFC at some point in the future.

Reply:
Good to know.  Thanks.

Reply:
My reading of that RFC is that X- prefixes are deprecated pretty much for all cases - hence the title _Deprecating the "X-" Prefix and Similar Constructs in Application Protocols_

If it's not too late to fix this before the beta, it'd be great to sort that out.